File: Baitap1/Complete/Java/7.4.2.java
import java.util.List;

public class Main {

    static class Animal {
        protected int age;

        public Animal(int age) {
            this.age = age;
        }
    }

    static class Person extends Animal {
        private String name;
        private List<String> friends;

        public Person(String name, int age, List<String> friends) {
            super(age);
            setName(name);
            this.friends = friends;
        }

        public Person(String name, int age) {
            this(name, age, null);
        }

        public void setName(String name) {
            this.name = name;
        }

        public String getName() {
            return name;
        }

        public List<String> getFriends() {
            return friends;
        }

        public int getAge() {
            return age;
        }

        public void speak() {
            System.out.println("Hello");
        }

        public void addFriend(String friendName) {
            if (!friends.contains(friendName))
                friends.add(friendName);
        }

        public void ageDiff(Person other) {
            int diff = age - other.age;
            System.out.println(Math.abs(diff) + " year(s) difference");
        }

        public String toString() {
            return "Person: " + name + ": " + Integer.toString(age);
        }
    }

    public static void main(String[] args) {
        System.out.println("\n---- Person tests ----");
        Person p1 = new Person("Jack", 30);
        Person p2 = new Person("Jill", 25);
        System.out.println(p1.getName());
        System.out.println(p1.getAge());
        System.out.println(p2.getName());
        System.out.println(p2.getAge());
        System.out.println(p1);
        p1.speak();
        p1.ageDiff(p2);
    }
}

File: Baitap1/Complete/Python/7.4.2.py
class Animal:
    def __init__(self, age):
        self.age = age

class Person(Animal):
    def __init__(self, name, age):
        Animal.__init__(self, age)
        self.set_name(name)
        self.friends = []

    def set_name(self, name):
        self.name = name

    def get_name(self):
        return self.name

    def get_friends(self):
        return self.friends

    def get_age(self):
        return self.age

    def speak(self):
        print("Hello")

    def add_friend(self, fname):
        if fname not in self.friends:
            self.friends.append(fname)

    def age_diff(self, other):
        diff = self.age - other.age
        print(abs(diff), "year(s) difference")

    def __str__(self):
        return "Person: " + str(self.name) + ": " + str(self.age)

print("\n---- Person tests ----")
p1 = Person("Jack", 30)
p2 = Person("Jill", 25)
print(p1.get_name())
print(p1.get_age())
print(p2.get_name())
print(p2.get_age())
print(p1)
p1.speak()
p1.age_diff(p2)
